exporter/redis.go

```diff
@@ -599,6 +629,98 @@ func doRedisCmd(c redis.Conn, cmd string, args ...interface{}) (reply interface{
 	return res, err
 }
 
+var errNotFound = errors.New("key not found")
+
+// getKeyInfo takes a key and returns the type, and the size or length of the value stored at that key.
+func getKeyInfo(c redis.Conn, key string) (info keyInfo, err error) {
+
+	keyType, err := redis.String(c.Do("TYPE", key))
+	if err != nil {
+		return info, err
+	}
+
+	info.keyType = keyType
+
+	switch keyType {
+	case "none":
+		return info, errNotFound
+	case "string":
+		if size, err := redis.Int64(c.Do("PFCOUNT", key)); err == nil {
+			info.keyType = "hyperloglog"
+			info.size = float64(size)
+		} else if size, err := redis.Int64(c.Do("STRLEN", key)); err == nil {
+			info.size = float64(size)
+		}
+	case "list":
+		if size, err := redis.Int64(c.Do("LLEN", key)); err == nil {
+			info.size = float64(size)
+		}
+	case "set":
+		if size, err := redis.Int64(c.Do("SCARD", key)); err == nil {
+			info.size = float64(size)
+		}
+	case "zset":
+		if size, err := redis.Int64(c.Do("ZCARD", key)); err == nil {
+			info.size = float64(size)
+		}
+	case "hash":
+		if size, err := redis.Int64(c.Do("HLEN", key)); err == nil {
+			info.size = float64(size)
+		}
+	default:
+		err = fmt.Errorf("Unknown type: %v for key: %v", keyType, key)
+	}
+
+	return info, err
+}
+
+// scanForKeys returns a list of keys matching `pattern` by using `SCAN`, which is safer for production systems than using `KEYS`.
+// This function was taken from: https://github.com/reisinger/examples-redigo
+func scanForKeys(c redis.Conn, pattern string) ([]string, error) {
+	iter := 0
+	keys := []string{}
+	for {
+		arr, err := redis.Values(c.Do("SCAN", iter, "MATCH", pattern))
+		if err != nil {
+			return keys, fmt.Errorf("error retrieving '%s' keys", pattern)
+		}
+
+		iter, _ = redis.Int(arr[0], nil)
+		k, _ := redis.Strings(arr[1], nil)
+		keys = append(keys, k...)
+
+		if iter == 0 {
+			break
+		}
+	}
+
+	return keys, nil
+}
+
+// getKeysFromPatterns does a SCAN for a key if the key contains pattern characters
+func getKeysFromPatterns(c redis.Conn, keys []dbKeyPair) (expandedKeys []dbKeyPair, err error) {
+	expandedKeys = []dbKeyPair{}
+	for _, k := range keys {
+		if regexp.MustCompile(`[\?\*\[\]\^]+`).MatchString(k.key) {
+			_, err := c.Do("SELECT", k.db)
+			if err != nil {
+				return expandedKeys, err
+			}
+			keyNames, err := scanForKeys(c, k.key)
+			if err != nil {
```

Comment:
why abort here and not check the remaining values in keys?

Reply:
Good point; changed.